### DISCUSSION TITLE
CI/CD enhancements

### DISCUSSION THREAD

**Discussion by mloning** [alan-turing-institute/sktime#1529] [Ideas]
A few useful change to improve our release/distribution cycle:

- [x] compile manylinux wheels (see #188)
- [x] set up CI to directly upload wheels to PyPI
- [x] include Makefile to automise release workflow (see this [intro](https://the-turing-way.netlify.com/make/make.html), also see e.g. https://github.com/alan-turing-institute/CleverCSV/blob/master/make_release.py)
- [x] enforce code formatting in released code (see #208, #256)
- [x] add coverage for unit tests (see #394)
- [ ] add code quality/climate checks (see https://lgtm.com, https://www.codacy.com, https://github.com/marketplace/code-climate, etc)
- [ ] consider [cibuildwheel](https://github.com/joerick/cibuildwheel/tree/master/cibuildwheel) package to simplify wheel building across platforms
- [ ] improve packaging for macOS wheels with [delocate](https://github.com/matthew-brett/delocate), also included in cibuildwheel, also see https://iscinumpy.gitlab.io/post/azure-devops-python-wheels/ 
- [x] distribute sktime via conda-forge: https://conda-forge.org/#contribute (see #326)
- [x] set up [pre-commit config](https://ljvmiranda921.github.io/notebook/2018/06/21/precommits-using-black-and-flake8/) (see #394)
- [x] add Black for code formatting (see #394)
- [x] add [check-manifest](https://pypi.org/project/check-manifest/) (see #394)
- [x] add [bugbear](https://pypi.org/project/flake8-bugbear/) to flake8 (see #394)
- [x] update setup.cfg, see e.g. https://github.com/scikit-hep/boost-histogram/blob/develop/setup.cfg (see #394)
- [ ] check out other [flake8 extensions](https://github.com/DmytroLitvinov/awesome-flake8-extensions)
- [ ] add checks for broken website links: https://fastpages.fast.ai/fastlinkcheck/
- [ ] use GitHub Action to check if Binder works and pre-build Docker images for quicker launch: https://github.com/jupyterhub/repo2docker-action#use-github-actions-to-cache-the-build-for-binderhub 
- [ ] explore [pre-commit-ci](https://pre-commit.ci) as suggested in #794 

#### Extra resources
* For building/distribution on GitHub Actions checkout: https://scikit-hep.org/developer/gha_wheels
* Useful links from [reproducible research lunch](https://github.com/alan-turing-institute/the-turing-way/issues/746)  
* scikit-learn's [release process](https://scikit-learn.org/dev/developers/maintainer.html) 
* Check out https://python-poetry.org thanks to @hiqbal2
* Consider [setuptools_scm](https://github.com/pypa/setuptools_scm) for versioning (software configuration management)

**Comment by xuyxu** [comment]
Hi @mloning, I would like to work on using `cibuildwheel` to simplify the workflow on package building. You can check our [package](https://github.com/LAMDA-NJU/Deep-Forest/blob/master/.github/workflows/build-wheels.yml) that uses cibuildwheel, it is quite convenient, and we can even build wheels for architectures like `aarch64`.

I have checked all supported python versions and platforms of sktime on [PyPI](https://pypi.org/project/sktime/#files), all of them can be built with cibuildwheel.

**EDIT**: Another benefit is that we can have sktime listed on the homepage of [cibuildwheel](https://github.com/joerick/cibuildwheel) as a way of advertisement ;-).

**Comment by mloning** [comment]
@xuyxu sounds good, one issue may be support for older macOS versions (see #748)

**Comment by TonyBagnall** [comment]
hi @mloning Im doing a tidy up, 200+ issues triggers me. Can this be closed?

**Reply by mloning** [reply]
Yes, turned it into a discussion now.

**Comment by fkiraly** [comment]
@freddyaboulton, @lmmentel, have you seen this?

**Reply by lmmentel** [reply]
Yes, I've seen it. Most of the concrete items are already implemented but in case someone feels that something is missing in CI it would be best to turn it into a specific issue so we can start working on it.
